# Employee Retention Analysis

Employee retention data is from "IBM HR Analytics Employee Attrition & Performance" from Kaggle:

https://www.kaggle.com/datasets/pavansubhasht/ibm-hr-analytics-attrition-dataset

Some of the data features include Education, Job Involvement and Satisfaction, Performance, Work life balance, etc... 

See data source for more detail

## Purpose

I will understand what features correlate with employee attrition. This information can be used to help predict people who may be planning on leaving there position. Once identified, the company can have a discussion with the employee to understand if there are things within their job and/or career path that they would like to see imporoved. If at risk employees can be identified early, then this will help both the company and employees. The employee can improve their job and life satisfaction with proper intervention. Also, the company can retain their top talent and minimize hiring and training costs.

## ETL

### Import Data and Load Libraries

In [2]:
import altair as alt
import pandas as pd
#from sklearn.preprocessing import OneHoteEncoder, OrdinalEncoder

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500) 


In [3]:
df = pd.read_csv('EmployeeRetention.csv')
typeDict = {'Age':'int16',
            'Attrition':'category',
            'BusinessTravel':'category',
            'DailyRate':'int32',
            'Department':'category',
            'DistanceFromHome':'int16',
            'Education':'category',
            'EducationField':'category',
            'EmployeeCount':'int32',
            'EmployeeNumber':'int32',
            'EnvironmentSatisfaction':'category',
            'Gender':'category',
            'HourlyRate':'int32',
            'JobInvolvement':'category',
            'JobLevel':'category',
            'JobRole':'category',
            'JobSatisfaction':'category',
            'MaritalStatus':'category',
            'MonthlyIncome':'int32',
            'MonthlyRate':'int32',
            'NumCompaniesWorked':'int16',
            'Over18':'bool',
            'OverTime':'category',
            'PercentSalaryHike':'int16',
            'PerformanceRating':'category',
            'RelationshipSatisfaction':'category',
            'StandardHours':'int16',
            'StockOptionLevel':'category',
            'TotalWorkingYears':'int16',
            'TrainingTimesLastYear':'int16',
            'WorkLifeBalance':'category',
            'YearsAtCompany':'int16',
            'YearsInCurrentRole':'int16',
            'YearsSinceLastPromotion':'int16',
            'YearsWithCurrManager':'int16'
            }
df = df.astype(typeDict)



## Change the features from integers to alphanumeric categories, prior to teaching a model, encorporate OrdinalEncoder
attrition = {'Yes': True,
             'No': False}

Over18 = {'Y': True,
          'N': False}

OverTime = {'Yes': True,
             'No': False}

education = {1:'High School',
             2:'College',
             3:'Bachelor',
             4:'Master',
             5:'Doctor'}

environmentSatisfaction = {1:'Low',
                           2:'Medium',
                           3:'High',
                           4:'Very High'}

jobInvolvement = {1:'Low',
                  2:'Medium',
                  3:'High',
                  4:'Very High'}

jobSatisfaction = {1:'Low',
                   2:'Medium',
                   3:'High',
                   4:'Very High'}

performanceRating = {1:'Low',
                     2:'Good',
                     3:'Excellent',
                     4:'Outstanding'}

relationshipSatisfaction = {1:'Low',
                            2:'Medium',
                            3:'High',
                            4:'Very High'}

workLifeBalance = {1:'Bad',
                   2:'Good',
                   3:'Better',
                   4:'Best'}


## Defining the order of the ordinal features
educationOrder = ['High School','College','Bachelor','Master','Doctor']
ordinalOrder = ['Low','Medium','High','Very High']
performanceOrder = ['Low','Good','Excellent','Outstanding']
workOrder = ['Bad','Good','Better','Best']
jobOrder = [1,2,3,4,5]
stockOrder = [0,1,2,3]

##
df['Attrition'] = df['Attrition'].map(attrition)
df['Over18'] = df['Over18'].map(attrition)
df['OverTime'] = df['OverTime'].map(attrition)
df['Education'] = df['Education'].map(education)
df['EnvironmentSatisfaction'] = df['EnvironmentSatisfaction'].map(environmentSatisfaction)
df['JobInvolvement'] = df['JobInvolvement'].map(jobInvolvement)
df['JobSatisfaction'] = df['JobSatisfaction'].map(jobSatisfaction)
df['PerformanceRating'] = df['PerformanceRating'].map(performanceRating)
df['RelationshipSatisfaction'] = df['RelationshipSatisfaction'].map(relationshipSatisfaction)
df['WorkLifeBalance'] = df['WorkLifeBalance'].map(workLifeBalance)

##
df['Attrition'] = df['Attrition'].astype('bool')
df['Over18'] = df['Over18'].astype('bool')
df['OverTime'] = df['OverTime'].astype('bool')
df['Education'] = df['Education'].astype(pd.api.types.CategoricalDtype(categories = educationOrder, ordered = True))
df['EnvironmentSatisfaction'] = df['EnvironmentSatisfaction'].astype(pd.api.types.CategoricalDtype(categories = ordinalOrder, ordered = True))
df['JobInvolvement'] = df['JobInvolvement'].astype(pd.api.types.CategoricalDtype(categories = ordinalOrder, ordered = True))
df['JobSatisfaction'] = df['JobSatisfaction'].astype(pd.api.types.CategoricalDtype(categories = ordinalOrder, ordered = True))
df['PerformanceRating'] = df['PerformanceRating'].astype(pd.api.types.CategoricalDtype(categories = performanceOrder, ordered = True))
df['RelationshipSatisfaction'] = df['RelationshipSatisfaction'].astype(pd.api.types.CategoricalDtype(categories = ordinalOrder, ordered = True))
df['WorkLifeBalance'] = df['WorkLifeBalance'].astype(pd.api.types.CategoricalDtype(categories = workOrder, ordered = True))
df['JobLevel'] = df['JobLevel'].astype(pd.api.types.CategoricalDtype(categories = jobOrder, ordered = True))
df['StockOptionLevel'] = df['StockOptionLevel'].astype(pd.api.types.CategoricalDtype(categories = stockOrder, ordered = True))


In [4]:
df.describe(include = 'all')

Age Attrition BusinessTravel    DailyRate  \
count   1470.000000      1470           1470  1470.000000   
unique          NaN         2              3          NaN   
top             NaN     False  Travel_Rarely          NaN   
freq            NaN      1233           1043          NaN   
mean      36.923810       NaN            NaN   802.485714   
std        9.135373       NaN            NaN   403.509100   
min       18.000000       NaN            NaN   102.000000   
25%       30.000000       NaN            NaN   465.000000   
50%       36.000000       NaN            NaN   802.000000   
75%       43.000000       NaN            NaN  1157.000000   
max       60.000000       NaN            NaN  1499.000000   

                    Department  DistanceFromHome Education EducationField  \
count                     1470       1470.000000      1470           1470   
unique                       3               NaN         5              6   
top     Research & Development               NaN  Bachelor  Life Sciences   
freq                       961               NaN       572            606   
mean                       NaN          9.192517       NaN            NaN   
std                        NaN          8.106864       NaN            NaN   
min                        NaN          1.000000       NaN            NaN   
25%                        NaN          2.000000       NaN            NaN   
50%                        NaN          7.000000       NaN            NaN   
75%                        NaN         14.000000       NaN            NaN   
max                        NaN         29.000000       NaN            NaN   

        EmployeeCount  EmployeeNumber EnvironmentSatisfaction Gender  \
count          1470.0     1470.000000                    1470   1470   
unique            NaN             NaN                       4      2   
top               NaN             NaN                    High   Male   
freq              NaN             NaN                     453    882   
mean              1.0     1024.865306                     NaN    NaN   
std               0.0      602.024335                     NaN    NaN   
min               1.0        1.000000                     NaN    NaN   
25%               1.0      491.250000                     NaN    NaN   
50%               1.0     1020.500000                     NaN    NaN   
75%               1.0     1555.750000                     NaN    NaN   
max               1.0     2068.000000                     NaN    NaN   

         HourlyRate JobInvolvement  JobLevel          JobRole JobSatisfaction  \
count   1470.000000           1470    1470.0             1470            1470   
unique          NaN              4       5.0                9               4   
top             NaN           High       1.0  Sales Executive       Very High   
freq            NaN            868     543.0              326             459   
mean      65.891156            NaN       NaN              NaN             NaN   
std       20.329428            NaN       NaN              NaN             NaN   
min       30.000000            NaN       NaN              NaN             NaN   
25%       48.000000            NaN       NaN              NaN             NaN   
50%       66.000000            NaN       NaN              NaN             NaN   
75%       83.750000            NaN       NaN              NaN             NaN   
max      100.000000            NaN       NaN              NaN             NaN   

       MaritalStatus  MonthlyIncome   MonthlyRate  NumCompaniesWorked Over18  \
count           1470    1470.000000   1470.000000         1470.000000   1470   
unique             3            NaN           NaN                 NaN      1   
top          Married            NaN           NaN                 NaN   True   
freq             673            NaN           NaN                 NaN   1470   
mean             NaN    6502.931293  14313.103401            2.693197    NaN   
std              NaN    4707.956783   71

In [5]:
df.value_counts('Attrition')

Attrition
False    1233
True      237
dtype: int64

#### Additional Cleaning from above information

**Things to note:**

- [ ] Labels are unbalanced (237 True: 1233 False)
- [ ] Is MonthlyIncome colinear with MonthlyRate
- [x] EmployeeCount features all have a value of 1. Removed
- [x] StandardHours features all have a value of 80. Removed

In [6]:
del df['EmployeeCount'] #All values were equal to 1
del df['StandardHours'] #All values were equal to 80

## Exploratory Data Analysis

In [19]:
continuous_field = list(df.select_dtypes(include='number').columns)

alt.Chart(df).mark_point().encode(
    alt.X(alt.repeat("column"), type='quantitative'),
    alt.Y(alt.repeat("row"), type='quantitative'),
    color = 'Attrition',
    opacity = alt.value(0.1)
).properties(
    width = 200,
    height = 200
).repeat(
    row = continuous_field,
    column = continuous_field[::-1]
)

alt.RepeatChart(...)

### Violin Plots

In [47]:
## VIOLIN PLOT
yAxis = 'Age'

alt.Chart(df).transform_density(
    yAxis,
    as_=[yAxis, 'density'],
    extent=[5, 50],
    groupby=['Attrition']
).mark_area(orient='horizontal').encode(
    y=str(yAxis+":Q"),
    color='Attrition:N',
    x=alt.X(
        'density:Q',
        stack='center',
        impute=None,
        title=None,
        axis=alt.Axis(labels=False, values=[0],grid=False, ticks=True),
    ),
    column=alt.Column(
        'Attrition:N',
        header=alt.Header(
            titleOrient='bottom',
            labelOrient='bottom',
            labelPadding=0,
        ),
    )
).properties(
    width=100
).configure_facet(
    spacing=0
).configure_view(
    stroke=None
).interactive()

alt.Chart(...)

### Swarm Plot

In [68]:
yAxis = 'Age'
alt.Chart(df).mark_circle(size=3).encode(
    x=alt.X(
        'jitter:Q',
        title = None,
        axis = alt.Axis(values=[0], ticks = True, grid = False, labels = False),
        scale = alt.Scale(),
    ),
    y = alt.Y(str(yAxis+":Q")),
    color = alt.Color('Attrition:N',legend=None),
    column = alt.Column(
        'Attrition:N',
        header = alt.Header(
            labelAngle = -90,
            titleOrient='top',
            labelOrient = 'bottom',
            labelAlign = 'right',
            labelPadding = 3,
        ),
    ),
).transform_calculate(
    jitter = 'sqrt(-1000*log(random()))*cos(2*PI*random())'
).configure_facet(
    spacing = 0
).configure_view(
    stroke = None
).properties(height = 300, width = 100
).interactive()

alt.Chart(...)

In [42]:
continuous_field


['Age',
 'DailyRate',
 'DistanceFromHome',
 'EmployeeNumber',
 'HourlyRate',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']